File: dubox-frontend/src/app/core/validators/meaningful-text.validator.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

/**
 * Validator to ensure text input meets minimum length requirement
 */
export class MeaningfulTextValidator {
  private static readonly MINIMUM_LENGTH = 5;

  /**
   * Validator function for Angular forms
   */
  static validate(): ValidatorFn {
    return (control: AbstractControl): ValidationErrors | null => {
      const value = control.value;

      if (!value) {
        return null; // Let required validator handle empty values
      }

      const result = MeaningfulTextValidator.isValid(value);
      
      if (!result.isValid) {
        return { meaningfulText: { message: result.errorMessage } };
      }

      return null;
    };
  }

  /**
   * Validates if the text meets minimum length requirement
   */
  static isValid(text: string | null | undefined): { isValid: boolean; errorMessage?: string } {
    // Check if null or empty
    if (!text || !text.trim()) {
      return {
        isValid: false,
        errorMessage: `This field is required and must contain at least ${this.MINIMUM_LENGTH} characters.`
      };
    }

    // Remove extra whitespace for validation
    const cleanedText = text.trim().replace(/\s+/g, ' ');

    // Check minimum length
    if (cleanedText.length < this.MINIMUM_LENGTH) {
      return {
        isValid: false,
        errorMessage: `This field must contain at least ${this.MINIMUM_LENGTH} characters. Current length: ${cleanedText.length} characters.`
      };
    }

    return { isValid: true };
  }

  /**
   * Gets the minimum required length
   */
  static getMinimumLength(): number {
    return this.MINIMUM_LENGTH;
  }
}

File: Dubox.Application/Validators/MeaningfulTextValidator.cs
using System.Text.RegularExpressions;

namespace Dubox.Application.Validators
{
    public static class MeaningfulTextValidator
    {
        private const int MinimumLength = 50;

        public static bool IsValid(string? text, out string errorMessage)
        {
            errorMessage = string.Empty;

            // Check if null or empty
            if (string.IsNullOrWhiteSpace(text))
            {
                errorMessage = $"This field is required and must contain at least {MinimumLength} characters.";
                return false;
            }

            // Remove extra whitespace for validation
            var cleanedText = Regex.Replace(text.Trim(), @"\s+", " ");

            // Check minimum length
            if (cleanedText.Length < MinimumLength)
            {
                errorMessage = $"This field must contain at least {MinimumLength} characters. Current length: {cleanedText.Length} characters.";
                return false;
            }

            return true;
        }

        /// <summary>
        /// Gets the minimum required length
        /// </summary>
        public static int GetMinimumLength() => MinimumLength;
    }
}
